File: native/kotlin/codegen/src/main/java/com/aws/blocks/kotlin/model/RpcModel.kt
package com.aws.blocks.kotlin.model

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

// ── Info Object Hierarchy ─────────────────────────────────────────

data class Contact(
    val name: String? = null,
    val url: String? = null,
    val email: String? = null,
)

data class License(
    val name: String,
    val url: String? = null,
)

data class Info(
    val title: String,
    val version: String,
    val description: String? = null,
    val termsOfService: String? = null,
    val contact: Contact? = null,
    val license: License? = null,
)

// ── Server Objects ────────────────────────────────────────────────

data class ServerVariable(
    val default: String,
    val description: String? = null,
    val enum: List<String> = emptyList(),
)

/** A single entry from the OpenRPC `servers` array. */
data class Server(
    val name: String,
    val url: String,
    val description: String? = null,
    val summary: String? = null,
    val variables: Map<String, ServerVariable> = emptyMap(),
)

// ── Method Metadata Classes ───────────────────────────────────────

data class ExternalDocumentation(
    val url: String,
    val description: String? = null,
)

data class Tag(
    val name: String,
    val summary: String? = null,
    val description: String? = null,
    val externalDocs: ExternalDocumentation? = null,
)

data class Link(
    val name: String,
    val summary: String? = null,
    val description: String? = null,
    val method: String? = null,
    val params: Map<String, Any>? = null,
    val server: Server? = null,
)

data class Error(
    val code: Int,
    val message: String,
    val data: TypeRef? = null,
)

data class ExamplePairing(
    val name: String,
    val summary: String? = null,
    val description: String? = null,
    val params: List<JsonElement> = emptyList(),
    val result: JsonElement? = null,
)

// ── ContentDescriptor (replaces Parameter) ────────────────────────

data class ContentDescriptor(
    val name: String,
    val schema: TypeRef,
    val description: String? = null,
    val summary: String? = null,
    val required: Boolean = false,
    val deprecated: Boolean = false,
)

// ── Field (modified: optional → required) ─────────────────────────

data class Field(
    val name: String,
    val type: TypeRef,
    val required: Boolean = false,
    val description: String? = null,
    val defaultValue: JsonElement? = null,
)

// ── Method (was MethodSignature) ──────────────────────────────────

data class Method(
    val name: String,
    val params: List<ContentDescriptor>,
    val result: ContentDescriptor? = null,
    val description: String? = null,
    val summary: String? = null,
    val tags: List<Tag> = emptyList(),
    val errors: List<Error> = emptyList(),
    val links: List<Link> = emptyList(),
    val examples: List<ExamplePairing> = emptyList(),
    val externalDocs: ExternalDocumentation? = null,
    val deprecated: Boolean = false,
    val paramStructure: String? = null,
)

// ── Root Model ────────────────────────────────────────────────────

/** Root model produced by the parser. */
data class RpcModel(
    val info: Info,
    val methods: List<Method>,
    val servers: List<Server> = emptyList(),
    val components: JsonObject? = null,
    val endpoint: String? = null,
)

/** A reference to a type, which can be primitive, inline, union, array, nullable, or imported. */
sealed interface TypeRef {
    data class Primitive(val tsType: String, val constraints: Constraints = Constraints.EMPTY) : TypeRef
    data class InlineObject(val fields: List<Field>, val additionalProperties: TypeRef? = null) : TypeRef
    data class UnionLiteral(val values: List<String>) : TypeRef
    data class ArrayType(val elementType: TypeRef, val constraints: Constraints = Constraints.EMPTY) : TypeRef
    data class MapType(val valueType: TypeRef) : TypeRef
    data class TupleType(val elements: List<TypeRef>) : TypeRef
    data class Nullable(val inner: TypeRef) : TypeRef
    data class Union(val members: List<TypeRef>) : TypeRef
    data class SchemaRef(val schemaName: String, val resolved: InlineObject, val description: String? = null) : TypeRef
    data class Transferable(val transferableName: String, val typeArgs: List<TypeRef>) : TypeRef
    data class ObjectWithOneOf(val fields: List<Field>, val oneOf: List<TypeRef>) : TypeRef
}

File: native/swift/Sources/BlocksCodegen/RPCModel.swift
import Foundation

// MARK: - Parser Output Model

/// A reference to a type as parsed from the spec. Not yet resolved.
indirect enum TypeRef {
    case primitive(kind: PrimitiveKind, constraints: Constraints = .empty)
    /// Object with named fields. When `additionalProperties` is non-nil the
    /// shape is open — the spec declared `T & Record<string, V>` so customers
    /// can pass arbitrary extra string-keyed values (e.g. Cognito custom
    /// attributes on sign-up). The value type is `V`.
    /// `embeddedUnion` carries an inner `oneOf` flattened at the top level —
    /// produced by the spec emitter's regroup pass for `{ action: K } & U[K]`
    /// shapes (e.g. Cognito's `confirmSignIn` with seven challenge bodies).
    /// The inner union shares the JSON envelope with the outer fields; codegen
    /// merges them at encode/decode time.
    case inlineObject(fields: [Field], additionalProperties: TypeRef?, embeddedUnion: TypeRef?)
    case unionLiteral(values: [String])
    case arrayType(elementType: TypeRef, constraints: Constraints = .empty)
    case nullable(inner: TypeRef)
    /// Sum type (`oneOf`). Discriminator detection happens later in the
    /// builder by inspecting members for shared single-literal fields.
    case union(members: [TypeRef])
    case schemaRef(name: String, resolved: TypeRef?)
    case transferable(blocksType: String, typeArgs: [TypeRef])
    /// `Record<string, T>` from JSON Schema's `additionalProperties` with no
    /// fixed properties. Renders as `[String: T]` in Swift.
    case mapType(valueType: TypeRef)
}

/// Native Swift primitives. `format`-specialized types (uuid, dateTime, url)
/// are NOT in this enum — they live as `ResolvedType.formattedType` so they
/// can carry their own constraint set and round-trip through Foundation's
/// dedicated parsers.
enum PrimitiveKind: String {
    case string, boolean, integer, number, void, unknown
}

struct Field {
    let name: String
    let type: TypeRef
    let required: Bool
    let description: String?
    /// `default` value from the spec, encoded as raw JSON.
    let defaultValue: String?

    init(name: String, type: TypeRef, required: Bool, description: String? = nil, defaultValue: String? = nil) {
        self.name = name
        self.type = type
        self.required = required
        self.description = description
        self.defaultValue = defaultValue
    }
}

struct ContentDescriptor {
    let name: String
    let schema: TypeRef
    let required: Bool
    let description: String?

    init(name: String, schema: TypeRef, required: Bool = false, description: String? = nil) {
        self.name = name
        self.schema = schema
        self.required = required
        self.description = description
    }
}

struct RPCMethod {
    let name: String
    let params: [ContentDescriptor]
    let result: ContentDescriptor?
    let description: String?

    init(name: String, params: [ContentDescriptor], result: ContentDescriptor?, description: String? = nil) {
        self.name = name
        self.params = params
        self.result = result
        self.description = description
    }
}

struct ServerDefinition {
    let name: String
    let url: String
}

public struct RPCModel {
    let methods: [RPCMethod]
    let servers: [ServerDefinition]
    let componentSchemas: [String: TypeRef]
}
